### ISSUE TITLE
Improve weight regularization configurability of optimizers.

### ISSUE THREAD

**Issue by veritas9872** [pytorch/pytorch#53718]
## 🚀 Feature
Allow the optimizer to decay different parameters of a model with variable weight decay rates with variable weight decay values.

## Motivation

Many recent papers, such as [NFNets](https://arxiv.org/abs/2102.06171), argue that weight decay should only be applied to weights of convolution and linear layers and not to normalization layers. Also, L2 weight decay is not the only weight decay that is being used in the community. Others may wish to use L1 weight decay or their own weight decay function. However, the current implementation of Adam, AdamW, etc. must use the L2 weight decay. Moreover, the `weight_decay` value must be a single floating point, with no official API to change the weight decay ratio as training progresses. A researcher may wish to decrease the weight decay as training progresses in order to restrain the model during the unstable initial phase of training while allowing it to find better values after it has reached a plateau.

Unlike Keras, where the weight decay function is set for each layer and parameter, the settings for weight decay must be specified in the optimizer in PyTorch. However, the current API means that it is very hard to specify which parameter should use which weight decay function with which ratio to combine with the loss function. 

To make matters worse, the implementation of optimizers in v1.8 has changed to C++ and CUDA low-level implementations. This means that there is no simple way even to set the per-parameter weight decay manually. The methods [here](https://discuss.pytorch.org/t/simple-l2-regularization/139) will not work for Adam and AdamW optimizers because the resulting weight decay will not be included in the proper locations.

As a result, users must implement their own versions of Adam and AdamW optimizers to implement weight decay that changes during training, as well as for different weight decay functions. Also, the per-parameter settings for giving parameter groups can be very inconvenient when, for example, trying to give only the weights of nn.Linear and nn.Conv* layers for weight decay and excluding all other layers (e.g. nn.GroupNorm) from weight decay. See [here](https://discuss.pytorch.org/t/weight-decay-only-for-weights-of-nn-linear-and-nn-conv/114348) for additional discussion.

## Pitch

Implement a new class of optim.weight_decay similar to optim.lr_scheduler that allows fine-tuned control of weight decay.

## Alternatives

Maybe some of the ideas in https://github.com/pytorch/pytorch/issues/22343 will be of relevance. However, I believe that the ideas presented here are broader as they also include per-parameter specifications.

## Additional context

I am aware that this is a big ask considering how much work will go into this. However, the current difficulty of configuring weight decay with optimizers in PyTorch may be hindering research into the importance of weight decay in deep learning and of other weight restriction schemes. I hope that this issue will at least raise awareness, even if it is not accepted.

cc @vincentqb

**Comment by veritas9872** [comment]
@izdeby 
It appears that this functionality is already available with TorchLayers.
https://github.com/szymonmaszke/torchlayers/blob/master/torchlayers/regularization.py
I would love to see this integrated into PyTorch, but I am closing this issue as this would require another request.
